File: python/binary-search-tree/binary_search_tree.py
class TreeNode(object):
    def __init__(self, value):
        self.value = value
        self.left = self.right = None


class BinarySearchTree(object):
    def __init__(self):
        self.root = None

    def add(self, value):
        node = TreeNode(value)

        if self.root is None:
            self.root = node
        else:
            parent = self.root
            while True:
                if value <= parent.value:
                    if parent.left is None:
                        parent.left = node
                        break
                    else:
                        parent = parent.left
                else:
                    if parent.right is None:
                        parent.right = node
                        break
                    else:
                        parent = parent.right

    def search(self, value):
        node = self.root
        while node is not None:
            if value == node.value:
                return node
            elif value < node.value:
                node = node.left
            else:
                node = node.right
        return None

    def list(self):
        result = []
        self.inorder(result, self.root)
        return result

    def inorder(self, result, node):
        if node is not None:
            self.inorder(result, node.left)
            result.append(node.value)
            self.inorder(result, node.right)

File: java/binary-search-tree/src/main/java/BinarySearchTree.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

public class BinarySearchTree<T extends Comparable<? super T>> {
  static class Node<T> {
    T data;
    Node<T> left;
    Node<T> right;

    Node(T data, Node<T> left, Node<T> right) {
      this.data = data;
      this.left = left;
      this.right = right;
    }

    T getData() {
      return data;
    }

    Node<T> getLeft() {
      return left;
    }

    Node<T> getRight() {
      return right;
    }
  }

  Node<T> root;

  void insert(T data) {
    Node<T> node = new Node<T>(data, null, null);
    if (root == null) {
      root = node;
    } else {
      Node<T> parent = root;
      while (true) {
        if (data.compareTo(parent.data) <= 0) {
          if (parent.left == null) {
            parent.left = node;

            break;
          } else {
            parent = parent.left;
          }
        } else {
          if (parent.right == null) {
            parent.right = node;

            break;
          } else {
            parent = parent.right;
          }
        }
      }
    }
  }

  Node<T> getRoot() {
    return root;
  }

  List<T> getAsLevelOrderList() {
    List<T> result = new ArrayList<>();
    Queue<Node<T>> queue = new ArrayDeque<>();
    if (root != null) {
      queue.offer(root);
    }
    while (!queue.isEmpty()) {
      Node<T> head = queue.poll();
      result.add(head.data);

      if (head.left != null) {
        queue.offer(head.left);
      }
      if (head.right != null) {
        queue.offer(head.right);
      }
    }

    return result;
  }

  List<T> getAsSortedList() {
    List<T> result = new ArrayList<>();
    inorder(result, root);

    return result;
  }

  void inorder(List<T> sorted, Node<T> node) {
    if (node != null) {
      inorder(sorted, node.left);
      sorted.add(node.data);
      inorder(sorted, node.right);
    }
  }
}
